File: app/src/main/java/com/expensio/domain/model/Expense.kt
package com.expensio.domain.model

data class Expense(
    val id: String,
    val groupId: String,
    val title: String,
    val amount: Double,
    val category: String,
    val paidByUserId: String?,
    val paidByGuestId: String?,
    val paidByName: String,
    val splitType: String,
    val createdAt: String,
)

data class ExpenseSplitDetail(
    val id: String,
    val userId: String?,
    val guestId: String?,
    val memberName: String,
    val amount: Double,
    val percentage: Double?,
)

data class ExpenseDetail(
    val expense: Expense,
    val splits: List<ExpenseSplitDetail>,
)

data class Balance(
    val userId: String?,
    val guestId: String?,
    val memberName: String,
    val netAmount: Double,
)

enum class SplitType { EQUAL, PERCENTAGE, EXACT }

val EXPENSE_CATEGORIES = listOf(
    "FOOD", "TRANSPORT", "SHOPPING", "ENTERTAINMENT",
    "HEALTH", "UTILITIES", "RENT", "OTHER"
)

File: backend/app/schemas/expense.py
import enum
import uuid
from datetime import datetime
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel


class SplitTypeEnum(str, enum.Enum):
    EQUAL = "EQUAL"
    PERCENTAGE = "PERCENTAGE"
    EXACT = "EXACT"


class SplitInput(BaseModel):
    user_id: Optional[uuid.UUID] = None
    guest_id: Optional[uuid.UUID] = None
    amount: Optional[Decimal] = None
    percentage: Optional[Decimal] = None


class ExpenseCreate(BaseModel):
    title: str
    amount: Decimal
    category: str = "OTHER"
    paid_by_user_id: Optional[uuid.UUID] = None
    paid_by_guest_id: Optional[uuid.UUID] = None
    split_type: SplitTypeEnum = SplitTypeEnum.EQUAL
    splits: list[SplitInput] = []


class ExpenseSplitResponse(BaseModel):
    id: uuid.UUID
    user_id: Optional[uuid.UUID]
    guest_id: Optional[uuid.UUID]
    member_name: str
    amount: Decimal
    percentage: Optional[Decimal]


class ExpenseResponse(BaseModel):
    id: uuid.UUID
    group_id: uuid.UUID
    title: str
    amount: Decimal
    category: str
    paid_by_user_id: Optional[uuid.UUID]
    paid_by_guest_id: Optional[uuid.UUID]
    paid_by_name: str
    split_type: str
    created_at: datetime

    model_config = {"from_attributes": True}


class ExpenseDetailResponse(ExpenseResponse):
    splits: list[ExpenseSplitResponse]


class BalanceResponse(BaseModel):
    user_id: Optional[uuid.UUID]
    guest_id: Optional[uuid.UUID]
    member_name: str
    net_amount: Decimal
